docs/enhancements/fep-002-config-hooks.md

```diff
@@ -0,0 +1,204 @@
+# EP: Definition of the Fleet and Device API spec
+<!-- this format is inspired by the K8S KEP format https://raw.githubusercontent.com/kubernetes/enhancements/master/keps/NNNN-kep-template/README.md -->
+## Summary
+
+This RFE defines a mechanism for per-device configuration extensiblity,
+where the device configuration can be instructed to talk to an external
+system or hook to fetch specific configurations for the device. This
+RFE also defines a talk-back mechanism for those hooks to tell flightctl
+that such configuration needs to be updated/revoked/etc.
+
+## Motivation
+
+Integration with external systems when devices may need specific
+per-device tokens, per-device configurations sometimes transient
+is a not-well covered use case in our API. Enabling such thing
+with today's API would not be possible. This is why we propose
+this configuration hooks API.
+
+This could be useful i.e. to integrate with ACM, where a service could
+be running on the ACM cluster, and the fleets configured to have
+the devices onboarded to ACM would point to that service in the fleet
+configuration items to fetch the configuration at specific
+points in time (hook points), anyway this is just an example, the
+mechanism should be generic to enable customers to leverage this.
+
+### Goals
+
+* Define hook points in the device configuration or flightctl
+  to enable the described functionality.
+
+* Allow an external system to be queried to retrieve per-device specific
+  configuration.
+
+* Allow the external system to call-back to flightctl to inform about
+  changes to be made in the previously supplied configuration.
+
+### Non-Goals
+
+* Define a specific product integration, this is a generic mechanism
+  that could be used with any external system.
+
+## Proposal
+
+We define a specific type of configuration that can be used to query an external system.
+
+```yaml
+apiVersion: v1alpha1
+kind: Fleet
+metadata:
+  creationTimestamp: "2024-04-30T14:06:17Z"
+  generation: 1
+  labels: {}
+  name: optical-inspector-production-fleet
+  owner: ResourceSync/basic-nginx-r3sourcesync
+spec:
+  selector:
+    matchLabels:
+      device: optical-inspector
+  template:
+    metadata:
+      generation: 1
+    spec:
+      config:
+...
+...
+      - configType: ExternalConfigProviderSpec
+        externalEndpoint:
+          config-name: my-config
+          url: https://acm-cluster-flightctl-hook.example.com/flightctl-hooks/
+          httpCredentials: my-credentials
```

Comment:
```suggestion
          httpCredentials: my-credentials
          hook-step: onFleetJoin
```

Reply:
Right, that could make sense, I would provide it as an optional field instead maybe?

```suggestion
          httpCredentials: my-credentials
          hook-events:
            - onFleetJoin
            - onMetadataUpdate
            - onDevicePoll
            
```

In this way, the administrator can optionally be precise on which events to send. If not provided we would post all.

Reply:
Who would send these events to whom as a result? E.g. it'd be the fleet-controller sending "onFleetJoin/Leave" events and the device-controller the "onMetadataUpdate" events (assuming this refers to device.metadata). And the receiver could well be the Repository task so it could query the source.

What is confusing is this seems to suggest you want to send the events to the external system directly, i.e. sending it via Kafka or via callback URLs to one ore more consumers. However, event notifications are informational; you do not / cannot expect the receiver to be available or take an action and respond (sync or async). Instead, we'd want to _query_ the external system, or? And then we're back to what the Repository task already does.

Reply:
That is a good point, if the system is not there to answer on the other side we may need to either retry, or queue, a messaging system would be more robust in a case like this and would avoid the need to retry on our side.